File: algorithms/2_medium/D/01144_decrease_elements_to_make_array_zigzag/greedy.c
int movesToMakeZigzag(const int *restrict nums, const int len) {
    if (nums == NULL || len <= 2) { return 0; }

    int odd = 0;
    int even = 0;

    for (int i = 1; i < len; i += 2) {
        const int prev = nums[i - 1];
        const int next = i + 1 < len ? nums[i + 1] : INT_MAX;
        const int target = MIN(prev, next) - 1;
        odd += MAX(nums[i] - target, 0);
    }

    for (int i = 0; i < len; i += 2) {
        const int prev = i - 1 >= 0  ? nums[i - 1] : INT_MAX;
        const int next = i + 1 < len ? nums[i + 1] : INT_MAX;
        const int target = MIN(prev, next) - 1;
        even += MAX(nums[i] - target, 0);
    }

    return MIN(odd, even);
}

File: algorithms/2_medium/D/01144_decrease_elements_to_make_array_zigzag/greedy.rs
impl Solution {
    pub fn moves_to_make_zigzag(nums: Vec<i32>) -> i32 {
        if nums.len() <= 2 { return 0; }

        let len = nums.len();
        let (mut odd, mut even) = (0_i32, 0_i32);

        for i in (1..len).step_by(2) {
            let prev = nums[i - 1];
            let next = *nums.get(i + 1).unwrap_or(&i32::MAX);
            let target = prev.min(next) - 1;
            odd += std::cmp::max(nums[i] - target, 0);
        }

        for i in (0..len).step_by(2) {
            let prev = *nums.get(i - 1).unwrap_or(&i32::MAX);
            let next = *nums.get(i + 1).unwrap_or(&i32::MAX);
            let target = prev.min(next) - 1;
            even += std::cmp::max(nums[i] - target, 0);
        }

        odd.min(even)
    }
}
